js/src/logger.ts

```diff
@@ -406,15 +413,144 @@ interface UserInfo {
   id: string;
 }
 
-export class Project {
-  name: string;
+interface RegisteredProject {
   id: string;
-  org_id: string;
+  name: string;
+}
 
-  constructor(name: string, id: string, org_id: string) {
+export class Project {
+  name?: string;
+  id?: string;
+
+  constructor({ name, id }: { name?: string; id?: string }) {
     this.name = name;
     this.id = id;
-    this.org_id = org_id;
+  }
+
+  async lazyInit(): Promise<RegisteredProject> {
+    if (this.id === undefined) {
+      const response = await _state
+        .apiConn()
+        .post_json("api/project/register", {
+          project_name: this.name || GLOBAL_PROJECT,
+          org_id: _state.orgId,
+        });
+      this.id = response.project.id;
+      if (this.name === undefined) {
+        this.name = response.project.name;
+      }
+    } else if (this.name === undefined) {
+      const response = await _state.apiConn().get_json("api/project", {
+        id: this.id,
+      });
+      this.name = response.name;
+    }
+
+    return { id: this.id!, name: this.name! };
+  }
+}
+
+export interface LogOptions {
+  is_async?: boolean;
+}
+
+export class Logger {
+  lazyLogin: () => Promise<void>;
```

Comment:
I don't think you can do async getters. Happy to change pattern as we evolve the lazy methods

Reply:
Ok, but I don't think we fixed anything, because the user can still access `this.lazyProject.id`, which will be undefined until they call some other method on the logger. Is this what we want?

In theory, we could have an `id: Promise<string>` member, which resolves to the actual ID when we have completed a lazy login. So people can call `await logger.id` (if you really want a getter member rather than a getter method). But in principle, getter methods are safer because users cannot modify them.

Reply:
Oh I see you meant a "getter" method like `public get id()`. You can just return a promise from the getter method to accomplish "async getter".

Reply:
Okay let's do that in a follow up